datanode/entities/position.go

```diff
@@ -60,27 +64,61 @@ type Position struct {
 	Adjustment              decimal.Decimal // what a party was missing which triggered loss socialization
 	TxHash                  TxHash
 	VegaTime                time.Time
+	// keep track of trades that haven't been settled as separate fields
+	// these will be zeroed out once we process settlement events
+	PendingOpenVolume              int64
+	PendingRealisedPnl             decimal.Decimal
+	PendingUnrealisedPnl           decimal.Decimal
+	PendingAverageEntryPrice       decimal.Decimal
+	PendingAverageEntryMarketPrice decimal.Decimal
```

Comment:
It feels slightly inelegant to track these separately from their non-pending counterparts, is there any reason we can't just update the 'actual' fields (I guess there probably is or else you wouldn't have done it :). It's just kind of conceptually to me, when a trade is done, your realised PnL has really changed - it's not pending.. the only thing that mark-to-market is really doing is giving you a price, changing your unrealised PnL, so maybe that's the only thing that should change when we get a mark-to-market event; the rest we can figure out from just looking at trades?

* obv. it's a bit more complicated when markets settle or distressed positions get closed

Reply:
We no longer mark to market after trades are made, so the new/change in position isn't yet settled, that only happens when your new position gets marked to market. By the time that happens, other trades may have changed the mark price, so MTM will take your old position, at the last mark price, mark the old position to market (new mark price), and then mark the new position to the new mark price to get to the PnL values.

We can't perform the normal MTM stuff on the existing fields because we kind of need to know the old position + old mark price when the settle events come in. These fields merely exist to update the position data and provide a _"best guess"_ approximation of the realised PnL when trades are made. We know the old mark price and trade price, that gives us some profit/loss, but the old volume, from last MTM settlement is still used in the MTM settlement in the core, so the transfers performed in the core would not match up, and the position API would not accurately reflect the balance transfers

Reply:
I get that we don't mark to market after trades any more - but if I bought one widget @ 100.. then a trade comes through that says I sold one for 101 - then I've really made some realised PnL at that point right, I don't need to wait until MtM time, I don't need any price to tell me I've made a profit of 1.. 

Or are you perhaps saying that even though my profit is 'locked in' in a sense - I won't get my margin back etc.. until the next MtM cycle and that is what makes it 'pending'?

The PositionSettlement event when it comes should have the exact same set of trades in that we got from watching the trades as they came in right? So it should be an exact match.

That said perhaps there's some wisdom in not mucking around with the existing logic as it's tried and tested. Though it feels like a bit of a kludge it's probably the best way!

Reply:
Oh, I see what you mean now. I suppose that, all things considered, we could probably use the trades here and continuously update (un)RealisedPnL and average entry (market) price using the trade events, and just call the `mtm` function on the position when we get the position settlement event.

For now, it might be useful to just check and see whether these pending fields match up with the values we get when we've processed the trades in the `UpdatePositionSettlement` function.
I'm just not sure what the impact of this might be when a market gets settled, or in case of loss socialisation. I'd need to look in to that a bit more. A lot of the PnL stuff (and MTM settlements for that matter) were implemented before we changed to periodic MTM settlements, so there's almost certainly room for improvement as a result

Reply:
Yes fair enough